src/h3lib/lib/iterators.c

```diff
@@ -0,0 +1,322 @@
+                                skip digit
+
+and so on.
+
+ */
+
+/**
+ * Initialize a IterCellsChildren struct representing the sequence giving
+ * the children of cell `h` at resolution `childRes`.
+ *
+ * At any point in the iteration, starting once
+ * the struct is initialized, IterCellsChildren.h gives the current child.
+ *
+ * Also, IterCellsChildren.h == H3_NULL when all the children have been iterated
+ * through, or if the input to `iterInitParent` was invalid.
+ */
+IterCellsChildren iterInitParent(H3Index h, int childRes) {
+    IterCellsChildren it;
+
+    it._parentRes = H3_GET_RESOLUTION(h);
+
+    if (childRes < it._parentRes || childRes > MAX_H3_RES || h == H3_NULL) {
+        return _null_iter();
+    }
+
+    it.h = _zeroIndexDigits(h, it._parentRes + 1, childRes);
+    H3_SET_RESOLUTION(it.h, childRes);
+
+    if (H3_EXPORT(isPentagon)(it.h)) {
+        // The skip digit skips `1` for pentagons.
+        // The "_skipDigit" moves to the left as we count up from the
+        // child resolution to the parent resolution.
+        it._skipDigit = childRes;
+    } else {
+        // if not a pentagon, we can ignore "skip digit" logic
+        it._skipDigit = -1;
+    }
+
+    return it;
+}
+
+/**
+ * Step a IterCellsChildren to the next child cell.
+ * When the iteration is over, IterCellsChildren.h will be H3_NULL.
+ * Handles iterating through hexagon and pentagon cells.
+ */
+void iterStepChild(IterCellsChildren* it) {
+    // once h == H3_NULL, the iterator returns an infinite sequence of H3_NULL
+    if (it->h == H3_NULL) return;
+
+    int childRes = H3_GET_RESOLUTION(it->h);
+
+    _inc(it, childRes);
+
+    for (int i = childRes; i >= it->_parentRes; i--) {
+        if (i == it->_parentRes) {
+            // if we're modifying the parent resolution digit, then we're done
+            *it = _null_iter();
+            return;
+        }
+
+        // PENTAGON_SKIPPED_DIGIT == 1
+        if (i == it->_skipDigit && _get(it, i) == PENTAGON_SKIPPED_DIGIT) {
+            // Then we are iterating through the children of a pentagon cell.
+            // All children of a pentagon have the property that the first
+            // nonzero digit between the parent and child resolutions is
+            // not 1.
+            // I.e., we never see a sequence like 00001.
+            // Thus, we skip the `1` in this digit.
+            _inc(it, i);
+            it->_skipDigit -= 1;
+            return;
+        }
+
+        // INVALID_DIGIT == 7
+        if (_get(it, i) == INVALID_DIGIT) {
+            _inc(it, i);  // zeros out it[i] and increments it[i-1] by 1
+        } else {
+            break;
+        }
+    }
+}
+
+// create iterator for children of base cell at given resolution
+IterCellsChildren iterInitBaseCellNum(int baseCellNum, int childRes) {
+    if (baseCellNum < 0 || baseCellNum >= NUM_BASE_CELLS || childRes < 0 ||
+        childRes > MAX_H3_RES) {
+        return _null_iter();
+    }
+
+    H3Index baseCell;
```

Comment:
Nit: This should probably be `res0Index`, we use `baseCell` for the `baseCellNum` most of the time